File: go/conformance_test.go
package smysl_test

// C-Read conformance, and the specification walked clause by clause.
//
// Two questions, as in the Python and JavaScript packages. Do we agree with the Rust —
// fixtures in, byte-identical bytes out. And do we do what the *document* says, section by
// section. A library can pass the first and fail the second.

import (
	"bytes"
	"encoding/binary"
	"math"
	"os"
	"path/filepath"
	"strings"
	"testing"

	smysl "github.com/vulogov/smysl/go"
)

const wireDir = "../fixtures/wire"
const specPath = "../Documentation/SMYSL_FORMAT_SPEC.md"

func fixtures(t *testing.T) []string {
	t.Helper()
	paths, err := filepath.Glob(filepath.Join(wireDir, "*.cbor"))
	if err != nil {
		t.Fatal(err)
	}
	if len(paths) == 0 {
		t.Fatal("no .cbor fixtures; the suite would pass vacuously")
	}
	return paths
}

func hex(t *testing.T, s string) []byte {
	t.Helper()
	out := make([]byte, 0, len(s)/2)
	for i := 0; i < len(s); i += 2 {
		var b byte
		for _, c := range s[i : i+2] {
			b <<= 4
			switch {
			case c >= '0' && c <= '9':
				b |= byte(c - '0')
			case c >= 'a' && c <= 'f':
				b |= byte(c-'a') + 10
			}
		}
		out = append(out, b)
	}
	return out
}

func TestSpecificationIsWhereWeThinkItIs(t *testing.T) {
	body, err := os.ReadFile(specPath)
	if err != nil {
		t.Fatalf("%s: %v", specPath, err)
	}
	if !strings.Contains(string(body), "Deterministic CBOR") {
		t.Fatal("the specification does not look like the specification")
	}
}

// -- C-Read: agreement with the reference ------------------------------------

func TestStoresReencodeByteIdentically(t *testing.T) {
	for _, path := range fixtures(t) {
		t.Run(filepath.Base(path), func(t *testing.T) {
			data, err := os.ReadFile(path)
			if err != nil {
				t.Fatal(err)
			}
			records, err := smysl.DecodeStore(data)
			if err != nil {
				t.Fatalf("decode: %v", err)
			}
			if len(records) == 0 {
				t.Fatal("decoded to nothing")
			}
			out, err := smysl.EncodeStore(records)
			if err != nil {
				t.Fatalf("encode: %v", err)
			}
			if !bytes.Equal(out, data) {
				for i := range data {
					if i >= len(out) || out[i] != data[i] {
						t.Fatalf("byte %d differs: original 0x%02x, re-encoded 0x%02x",
							i, data[i], out[i])
					}
				}
				t.Fatalf("length differs: original %d, re-encoded %d", len(data), len(out))
			}
			// Record by record too, so a compensating pair of errors cannot hide.
			for n, r := range records {
				again, err := r.Reencode()
				if err != nil {
					t.Fatalf("record %d (%s): %v", n, r.Name(), err)
				}
				if !bytes.Equal(again, r.Raw) {
					t.Fatalf("record %d (%s) changed", n, r.Name())
				}
			}
		})
	}
}

// -- §2.2  What is hashed -----------------------------------------------------

func TestUnitCoreKeysMatchTheTable(t *testing.T) {
	want := map[uint64]string{
		0: "schema", 1: "gist", 2: "body", 3: "detail", 4: "deps",
		5: "grounds", 6: "status", 7: "source", 8: "payload",
	}
	if len(smysl.UnitKeys) != len(want) {
		t.Fatalf("key count: got %d, want %d", len(smysl.UnitKeys), len(want))
	}
	for k, v := range want {
		if smysl.UnitKeys[k] != v {
			t.Errorf("key %d: got %q, want %q", k, smysl.UnitKeys[k], v)
		}
	}
}

func TestUnknownKeysAtNineAndAboveSurvive(t *testing.T) {
	body := &smysl.Map{Entries: []smysl.Pair{
		{Key: uint64(0), Value: "claim"},
		{Key: uint64(1), Value: "a gist"},
		{Key: uint64(6), Value: uint64(1)},
		{Key: uint64(9), Value: "a later addition"},
	}}
	raw, err := smysl.EncodeOne([]any{uint64(1), body})
	if err != nil {
		t.Fatal(err)
	}
	records, err := smysl.DecodeStore(raw)
	if err != nil {
		t.Fatal(err)
	}
	again, _ := records[0].Reencode()
	if !bytes.Equal(again, raw) {
		t.Fatal("an unknown key did not survive a round trip")
	}
	m := records[0].Body.(*smysl.Map)
	if v, ok := m.Get(uint64(9)); !ok || v != "a later addition" {
		t.Fatalf("unknown key 9 lost: %v", v)
	}
}

// -- §3  Deterministic CBOR, one per constraint -------------------------------

func TestForbiddenEncodingsAreRejected(t *testing.T) {
	cases := map[string]string{
		"non-shortest one-byte":   "1817",
		"non-shortest two-byte":   "1900ff",
		"non-shortest four-byte":  "1a0000ffff",
		"indefinite-length array": "9fff",
		"indefinite-length map":   "bfff",
		"indefinite-length bytes": "5fff",
		"null":                    "f6",
		"binary64 float":          "fb3ff0000000000000",
		"map keys out of order":   "a201000000",
		"duplicate map key":       "a200000000",
		"a tag (constraint 8)":    "c000",
	}
	for why, encoded := range cases {
		if _, _, err := smysl.DecodeOne(hex(t, encoded)); err == nil {
			t.Errorf("%s was accepted", why)
		}
	}
}

func TestConstraintTwoDoesNotApplyToFloatPayloads(t *testing.T) {
	// The clause the two earlier implementations had to guess at, now stated in §3.
	v, _, err := smysl.DecodeOne(hex(t, "fa3f800000"))
	if err != nil {
		t.Fatalf("1.0 was rejected: %v", err)
	}
	if v != float32(1.0) {
		t.Fatalf("got %v, want 1.0", v)
	}
}

func TestFloatsAreBinary32OnTheQuantisationGrid(t *testing.T) {
	f32 := func(v float32) []byte {
		out := []byte{0xfa}
		return binary.BigEndian.AppendUint32(out, math.Float32bits(v))
	}
	for _, ok := range []float32{0, 0.5, 1.0 / 1024, -3.25} {
		if _, _, err := smysl.DecodeOne(f32(ok)); err != nil {
			t.Errorf("%v was rejected: %v", ok, err)
		}
	}
	for _, bad := range []float32{0.1, 1.0 / 3} {
		if _, _, err := smysl.DecodeOne(f32(bad)); err == nil {
			t.Errorf("%v was accepted", bad)
		}
	}
	if _, _, err := smysl.DecodeOne(f32(float32(math.Inf(1)))); err == nil {
		t.Error("infinity was accepted")
	}
}

func TestNfcTextIsRequired(t *testing.T) {
	// Written as explicit bytes rather than as source literals. The first attempt used two
	// "é" characters in the source and an editor normalised them to the same one, so the
	// test asserted that a string differs from itself and reported the decoder as broken.
	// A test whose two inputs are secretly equal is worse than no test.
	decomposed := []byte{0x65, 0xcc, 0x81} // e + U+0301
	if _, _, err := smysl.DecodeOne(append([]byte{0x60 | byte(len(decomposed))}, decomposed...)); err == nil {
		t.Error("non-NFC text was accepted")
	}
	composed := []byte{0xc3, 0xa9} // U+00E9
	v, _, err := smysl.DecodeOne(append([]byte{0x60 | byte(len(composed))}, composed...))
	if err != nil {
		t.Errorf("composed text was rejected: %v", err)
	}
	if v != "\u00e9" {
		t.Errorf("got %q", v)
	}
}

func TestNestingIsBoundedAt128(t *testing.T) {
	if smysl.MaxNesting != 128 {
		t.Fatalf("§3 constraint 9 names 128, not %d", smysl.MaxNesting)
	}
	shallow := append(bytes.Repeat([]byte{0x81}, 100), 0x00)
	if _, _, err := smysl.DecodeOne(shallow); err != nil {
		t.Errorf("100 levels was rejected: %v", err)
	}
	deep := append(bytes.Repeat([]byte{0x81}, 200), 0x00)
	if _, _, err := smysl.DecodeOne(deep); err == nil {
		t.Error("200 levels was accepted")
	}
}

// -- §3.1  Record framing -----------------------------------------------------

func TestRecordTypeCodesMatchTheTable(t *testing.T) {
	want := map[uint64]string{
		1: "unit", 2: "attestation", 3: "relation", 4: "thread", 5: "view",
		6: "contention", 7: "pack_info", 8: "schema_decl", 9: "checkpoint", 10: "label_binding",
	}
	if len(smysl.RecordNames) != len(want) {
		t.Fatalf("code count: got %d, want %d", len(smysl.RecordNames), len(want))
	}
	for k, v := range want {
		if smysl.RecordNames[k] != v {
			t.Errorf("code %d: got %q, want %q", k, smysl.RecordNames[k], v)
		}
	}
}

func TestARecordIsATwoElementArray(t *testing.T) {
	for _, bad := range []string{"8301a000", "a101a0"} {
		if _, err := smysl.DecodeStore(hex(t, bad)); err == nil {
			t.Errorf("%s was accepted as a record", bad)
		}
	}
}

func TestUnknownRecordTypesArePreservedNotRejected(t *testing.T) {
	body := &smysl.Map{Entries: []smysl.Pair{{Key: uint64(0), Value: "a later shape"}}}
	raw, err := smysl.EncodeOne([]any{uint64(99), body})
	if err != nil {
		t.Fatal(err)
	}
	records, err := smysl.DecodeStore(raw)
	if err != nil {
		t.Fatalf("an unknown record type was rejected: %v", err)
	}
	if records[0].IsKnown() || records[0].Name() != "unknown(99)" {
		t.Fatalf("got %s", records[0].Name())
	}
	again, _ := records[0].Reencode()
	if !bytes.Equal(again, raw) {
		t.Fatal("an unknown record did not survive verbatim")
	}
}

func TestAnUnknownRecordBodyIsStillParsedStrictly(t *testing.T) {
	// §3.1: an unknown record cannot smuggle in a non-deterministic encoding.
	if _, err := smysl.DecodeStore(hex(t, "8218639fff")); err == nil {
		t.Error("an indefinite-length body inside an unknown record was accepted")
	}
}

func TestAStoreIsAConcatenationWithNoEnvelope(t *testing.T) {
	mk := func(gist string) []byte {
		body := &smysl.Map{Entries: []smysl.Pair{
			{Key: uint64(0), Value: "claim"},
			{Key: uint64(1), Value: gist},
			{Key: uint64(6), Value: uint64(1)},
		}}
		raw, err := smysl.EncodeOne([]any{uint64(1), body})
		if err != nil {
			t.Fatal(err)
		}
		return raw
	}
	joined := append(mk("first"), mk("second")...)
	records, err := smysl.DecodeStore(joined)
	if err != nil {
		t.Fatal(err)
	}
	if len(records) != 2 {
		t.Fatalf("got %d records, want 2", len(records))
	}
	for i, want := range []string{"first", "second"} {
		if v, ok := records[i].UnitField("gist"); !ok || v != want {
			t.Errorf("record %d gist: got %v, want %q", i, v, want)
		}
	}
	out, _ := smysl.EncodeStore(records)
	if !bytes.Equal(out, joined) {
		t.Error("a concatenated store did not re-encode to itself")
	}
}

func TestATruncatedTrailingRecordIsAnError(t *testing.T) {
	body := &smysl.Map{Entries: []smysl.Pair{{Key: uint64(1), Value: "g"}}}
	good, _ := smysl.EncodeOne([]any{uint64(1), body})
	if _, err := smysl.DecodeStore(append(good, 0x82, 0x01)); err == nil {
		t.Error("a truncated trailing record was ignored rather than reported")
	}
}

// -- §7  Conformance ----------------------------------------------------------

func TestWhatCReadCannotCheck(t *testing.T) {
	// Kept as a list rather than a silence. The largest entry is §2.3 — status is part of
	// identity — the paragraph the whole format rests on, which needs uids and so C-Produce.
	unreached := map[string]string{
		"§2.1 uid derivation":                "needs BLAKE3; C-Produce, not C-Read",
		"§2.3 status is part of identity":    "follows from uid derivation",
		"§4 canonical surface form":          "surface syntax is not decoded here",
		"§6 rules M, T, L, R, U, I, S, V, D": "semantic; C-Consume and above",
	}
	if len(unreached) == 0 {
		t.Fatal("if this is ever empty, say which class was implemented instead")
	}
	if _, ok := unreached["§2.3 status is part of identity"]; !ok {
		t.Error("the format's central claim dropped off the list of what is untested")
	}
}

File: nodejs/test/conformance.test.js
// C-Read conformance, and the specification walked clause by clause.
//
// Two questions, as in the Python package. Do we agree with the Rust — fixtures in,
// byte-identical bytes out. And do we do what the *document* says, section by section. A
// library can pass the first and fail the second, which is why both are here.
//
// `node --test`, no dependencies. A dependency doing part of the work would weaken the
// evidence this exists to provide.

import { test } from "node:test";
import assert from "node:assert/strict";
import { readFileSync, readdirSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { dirname, join } from "node:path";

import * as smysl from "../src/index.js";

const here = dirname(fileURLToPath(import.meta.url));
const WIRE = join(here, "..", "..", "fixtures", "wire");
const SPEC = join(here, "..", "..", "Documentation", "SMYSL_FORMAT_SPEC.md");
const fixtures = readdirSync(WIRE).filter((f) => f.endsWith(".cbor")).sort();

const bytes = (name) => new Uint8Array(readFileSync(join(WIRE, name)));
const hex = (s) => Uint8Array.from(s.match(/../g).map((b) => parseInt(b, 16)));

test("there are fixtures to read", () => {
  assert.ok(fixtures.length, "no .cbor fixtures; the suite would pass vacuously");
});

test("the specification is where we think it is", () => {
  assert.match(readFileSync(SPEC, "utf8"), /Deterministic CBOR/);
});

// -- C-Read: agreement with the reference ------------------------------------

for (const name of fixtures) {
  test(`${name} decodes`, () => {
    assert.ok(smysl.decodeStore(bytes(name)).length, `${name} decoded to nothing`);
  });

  test(`${name} re-encodes byte-identically`, () => {
    const data = bytes(name);
    const out = smysl.encodeStore(smysl.decodeStore(data));
    for (let i = 0; i < Math.min(data.length, out.length); i++) {
      assert.equal(out[i], data[i], `byte ${i} differs in ${name}`);
    }
    assert.equal(out.length, data.length, `length differs in ${name}`);
  });

  test(`${name}: every record re-encodes individually`, () => {
    smysl.decodeStore(bytes(name)).forEach((r, n) => {
      assert.deepEqual(r.reencode(), r.raw, `${name}: record ${n} (${r.name}) changed`);
    });
  });
}

// -- §2.2  What is hashed -----------------------------------------------------

test("§2.2 unit core keys match the table", () => {
  assert.deepEqual(
    [...smysl.UNIT_KEYS.entries()],
    [[0, "schema"], [1, "gist"], [2, "body"], [3, "detail"], [4, "deps"],
     [5, "grounds"], [6, "status"], [7, "source"], [8, "payload"]],
  );
});

test("§2.2/§5 keys at nine and above are unknown and kept", () => {
  const body = new Map([[0, "claim"], [1, "a gist"], [6, 1], [9, "a later addition"]]);
  const raw = smysl.encodeOne([1, body]);
  const [rec] = smysl.decodeStore(raw);
  assert.deepEqual(rec.reencode(), raw);
  assert.equal(rec.unitFields().get(9), "a later addition");
});

test("§2.2 an absent optional is omitted, not null", () => {
  assert.throws(() => smysl.decodeOne(hex("a102f6")), smysl.CborError);
});

// -- §3  Deterministic CBOR, one per constraint -------------------------------

for (const [data, why] of [
  ["1817", "non-shortest one-byte"],
  ["1900ff", "non-shortest two-byte"],
  ["1a0000ffff", "non-shortest four-byte"],
]) {
  test(`§3 constraint 2 rejects a ${why} integer`, () => {
    assert.throws(() => smysl.decodeOne(hex(data)), smysl.CborError);
  });
}

test("§3 constraint 2 does not apply to float payloads", () => {
  // 0x3F800000 is 1.0, not an over-long encoding of 1065353216.
  assert.equal(smysl.decodeOne(hex("fa3f800000")).value, 1.0);
});

for (const [data, why] of [["9fff", "array"], ["bfff", "map"], ["5fff", "bytes"], ["7fff", "text"]]) {
  test(`§3 constraint 3 rejects an indefinite-length ${why}`, () => {
    assert.throws(() => smysl.decodeOne(hex(data)), smysl.CborError);
  });
}

test("§3 constraint 4 requires ascending, non-duplicate keys", () => {
  smysl.decodeOne(hex("a20000010 0".replace(/ /g, "")));
  assert.throws(() => smysl.decodeOne(hex("a201000000")), smysl.CborError);
  assert.throws(() => smysl.decodeOne(hex("a200000000")), smysl.CborError);
});

test("§3 constraint 5 rejects null", () => {
  assert.throws(() => smysl.decodeOne(hex("f6")), smysl.CborError);
});

test("§3 constraint 6 requires NFC text", () => {
  const decomposed = new TextEncoder().encode("é");
  assert.throws(
    () => smysl.decodeOne(Uint8Array.from([0x60 | decomposed.length, ...decomposed])),
    smysl.CborError,
  );
  const composed = new TextEncoder().encode("é");
  assert.equal(
    smysl.decodeOne(Uint8Array.from([0x60 | composed.length, ...composed])).value,
    "é",
  );
});

test("§3 constraint 7 requires binary32 quantised to 1/1024", () => {
  const f32 = (v) => {
    const b = new ArrayBuffer(4);
    new DataView(b).setFloat32(0, v);
    return Uint8Array.from([0xfa, ...new Uint8Array(b)]);
  };
  for (const ok of [0.0, 0.5, 1 / 1024, -3.25]) {
    assert.equal(smysl.decodeOne(f32(ok)).value, ok);
  }
  assert.throws(() => smysl.decodeOne(f32(0.1)), smysl.CborError);
  assert.throws(() => smysl.decodeOne(f32(Infinity)), smysl.CborError);
  assert.throws(() => smysl.decodeOne(hex("fb3ff0000000000000")), smysl.CborError); // binary64
});

test("§3 constraint 9 bounds nesting at 128", () => {
  assert.equal(smysl.MAX_NESTING, 128);
  smysl.decodeOne(Uint8Array.from([...Array(100).fill(0x81), 0x00]));
  assert.throws(
    () => smysl.decodeOne(Uint8Array.from([...Array(200).fill(0x81), 0x00])),
    smysl.CborError,
  );
});

test("§3 rejects a tag (major type 6)", () => {
  assert.throws(() => smysl.decodeOne(hex("c000")), smysl.CborError);
});

// -- §3.1  Record framing -----------------------------------------------------

test("§3.1 record type codes match the table", () => {
  assert.deepEqual(
    [...smysl.RECORD_NAMES.entries()],
    [[1, "unit"], [2, "attestation"], [3, "relation"], [4, "thread"], [5, "view"],
     [6, "contention"], [7, "pack_info"], [8, "schema_decl"], [9, "checkpoint"],
     [10, "label_binding"]],
  );
});

test("§3.1 a record is a two-element array", () => {
  assert.throws(() => smysl.decodeStore(hex("8301a000")), smysl.CborError);
  assert.throws(() => smysl.decodeStore(hex("a101a0")), smysl.CborError);
});

test("§3.1 an unknown type code is preserved and skipped, not rejected", () => {
  const raw = smysl.encodeOne([99, new Map([[0, "a later shape"]])]);
  const [rec] = smysl.decodeStore(raw);
  assert.equal(rec.isKnown, false);
  assert.equal(rec.name, "unknown(99)");
  assert.deepEqual(rec.reencode(), raw);
});

test("§3.1 an unknown record's body is still parsed strictly", () => {
  assert.throws(() => smysl.decodeStore(hex("8218639fff")), smysl.CborError);
});

test("§3.1 a store is a concatenation with no envelope", () => {
  const a = smysl.encodeOne([1, new Map([[0, "claim"], [1, "first"], [6, 1]])]);
  const b = smysl.encodeOne([1, new Map([[0, "claim"], [1, "second"], [6, 1]])]);
  const joined = Uint8Array.from([...a, ...b]);
  const recs = smysl.decodeStore(joined);
  assert.deepEqual(recs.map((r) => r.unitFields().get("gist")), ["first", "second"]);
  assert.deepEqual(smysl.encodeStore(recs), joined);
});

test("§3.1 a truncated trailing record is an error, not a silent truncation", () => {
  const good = smysl.encodeOne([1, new Map([[0, "claim"], [1, "g"], [6, 1]])]);
  assert.throws(() => smysl.decodeStore(Uint8Array.from([...good, 0x82, 0x01])), smysl.CborError);
});

// -- §7  Conformance ----------------------------------------------------------

test("§7 what C-Read provably cannot check, kept as a list", () => {
  // Emptying this should be a decision, not an accident. The largest entry is §2.3 — status
  // is part of identity — the paragraph the whole format rests on, which needs uids and
  // therefore C-Produce.
  const unreached = {
    "§2.1 uid derivation": "needs BLAKE3; C-Produce, not C-Read",
    "§2.3 status is part of identity": "follows from uid derivation",
    "§4 canonical surface form": "surface syntax is not decoded here",
    "§6 rules M, T, L, R, U, I, S, V, D": "semantic; C-Consume and above",
  };
  assert.ok(Object.keys(unreached).length);
  assert.ok("§2.3 status is part of identity" in unreached);
});
